In [1]:
import torch
import torch.nn as nn
import torch.nn.functional as F
from torchsummary import summary
from tqdm import tqdm

from model1 import Net
from train_test_loop import *
from data_utils import *

/home/raghu/.local/lib/python3.10/site-packages/transformers/utils/generic.py:441: UserWarning: torch.utils._pytree._register_pytree_node is deprecated. Please use torch.utils._pytree.register_pytree_node instead.
  _torch_pytree._register_pytree_node(


In [2]:
torch.manual_seed(1)

In [3]:
BATCH_SIZE=128
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model = Net().to(device)
summary(model, input_size=(1, 28, 28))

----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1           [-1, 16, 26, 26]             144
              ReLU-2           [-1, 16, 26, 26]               0
       BatchNorm2d-3           [-1, 16, 26, 26]              32
           Dropout-4           [-1, 16, 26, 26]               0
            Conv2d-5           [-1, 32, 24, 24]           4,608
              ReLU-6           [-1, 32, 24, 24]               0
       BatchNorm2d-7           [-1, 32, 24, 24]              64
           Dropout-8           [-1, 32, 24, 24]               0
            Conv2d-9           [-1, 10, 24, 24]             320
        MaxPool2d-10           [-1, 10, 12, 12]               0
           Conv2d-11           [-1, 16, 10, 10]           1,440
             ReLU-12           [-1, 16, 10, 10]               0
      BatchNorm2d-13           [-1, 16, 10, 10]              32
          Dropout-14           [-1, 16,

In [4]:
default_train_transforms = transforms.Compose([
                                        transforms.RandomRotation((-7.0, 7.0), fill=(1,)),
                                       transforms.ToTensor(),
                                       transforms.Normalize((0.1307,), (0.3081,))])
default_test_transforms = transforms.Compose([
                        transforms.ToTensor(),
                        transforms.Normalize((0.1307,), (0.3081,))
                    ])

In [5]:
train_dataloader, test_dataloader = get_augmented_MNIST_dataset("~/work/data/", 
                                    train_tfms=default_train_transforms, test_tfms=default_test_transforms,
                                    batch_sz=BATCH_SIZE)

In [6]:
optimizer = torch.optim.SGD(model.parameters(), lr=0.01, momentum=0.9)
scheduler = get_schedulder(optimizer, "StepLR", step_size=6, gamma=0.5)

In [7]:
EPOCHS=15
for i in range(EPOCHS):
    print("Running EPOCH: "+str(i+1))
    train(model, device, train_dataloader, optimizer=optimizer, epoch=i)
    test(model, device, test_dataloader)

Running EPOCH: 1


Loss=0.11275836825370789 Batch_id=468 Accuracy=88.15: 100%|██████████████████████████| 469/469 [00:02<00:00, 207.98it/s]



Test set: Average loss: 0.0728, Accuracy: 9789/10000 (97.89%)

Running EPOCH: 2


Loss=0.042526815086603165 Batch_id=468 Accuracy=97.56: 100%|█████████████████████████| 469/469 [00:01<00:00, 238.37it/s]



Test set: Average loss: 0.0379, Accuracy: 9889/10000 (98.89%)

Running EPOCH: 3


Loss=0.14141039550304413 Batch_id=468 Accuracy=98.10: 100%|██████████████████████████| 469/469 [00:01<00:00, 246.84it/s]



Test set: Average loss: 0.0309, Accuracy: 9917/10000 (99.17%)

Running EPOCH: 4


Loss=0.04536054655909538 Batch_id=468 Accuracy=98.37: 100%|██████████████████████████| 469/469 [00:01<00:00, 236.72it/s]



Test set: Average loss: 0.0277, Accuracy: 9907/10000 (99.07%)

Running EPOCH: 5


Loss=0.1048704981803894 Batch_id=468 Accuracy=98.50: 100%|███████████████████████████| 469/469 [00:01<00:00, 239.91it/s]



Test set: Average loss: 0.0268, Accuracy: 9917/10000 (99.17%)

Running EPOCH: 6


Loss=0.09229431301355362 Batch_id=468 Accuracy=98.57: 100%|██████████████████████████| 469/469 [00:01<00:00, 236.74it/s]



Test set: Average loss: 0.0236, Accuracy: 9927/10000 (99.27%)

Running EPOCH: 7


Loss=0.026816047728061676 Batch_id=468 Accuracy=98.73: 100%|█████████████████████████| 469/469 [00:02<00:00, 232.69it/s]



Test set: Average loss: 0.0276, Accuracy: 9917/10000 (99.17%)

Running EPOCH: 8


Loss=0.028934000059962273 Batch_id=468 Accuracy=98.71: 100%|█████████████████████████| 469/469 [00:02<00:00, 224.44it/s]



Test set: Average loss: 0.0237, Accuracy: 9930/10000 (99.30%)

Running EPOCH: 9


Loss=0.028496213257312775 Batch_id=468 Accuracy=98.79: 100%|█████████████████████████| 469/469 [00:02<00:00, 231.52it/s]



Test set: Average loss: 0.0217, Accuracy: 9932/10000 (99.32%)

Running EPOCH: 10


Loss=0.023397324606776237 Batch_id=468 Accuracy=98.86: 100%|█████████████████████████| 469/469 [00:02<00:00, 225.35it/s]



Test set: Average loss: 0.0211, Accuracy: 9938/10000 (99.38%)

Running EPOCH: 11


Loss=0.04109947010874748 Batch_id=468 Accuracy=98.90: 100%|██████████████████████████| 469/469 [00:02<00:00, 228.92it/s]



Test set: Average loss: 0.0216, Accuracy: 9933/10000 (99.33%)

Running EPOCH: 12


Loss=0.07317029684782028 Batch_id=468 Accuracy=98.92: 100%|██████████████████████████| 469/469 [00:01<00:00, 235.30it/s]



Test set: Average loss: 0.0204, Accuracy: 9933/10000 (99.33%)

Running EPOCH: 13


Loss=0.006672200281172991 Batch_id=468 Accuracy=98.94: 100%|█████████████████████████| 469/469 [00:02<00:00, 227.29it/s]



Test set: Average loss: 0.0179, Accuracy: 9948/10000 (99.48%)

Running EPOCH: 14


Loss=0.07240815460681915 Batch_id=468 Accuracy=98.93: 100%|██████████████████████████| 469/469 [00:01<00:00, 236.60it/s]



Test set: Average loss: 0.0203, Accuracy: 9936/10000 (99.36%)

Running EPOCH: 15


Loss=0.07996340841054916 Batch_id=468 Accuracy=99.04: 100%|██████████████████████████| 469/469 [00:02<00:00, 234.41it/s]



Test set: Average loss: 0.0181, Accuracy: 9945/10000 (99.45%)



#### Target

Check the accuracy and size of the model discussed in class

#### Result:

99.45% accuracy but not consistent, model size 13808

#### Analysis:

Step 1 is a naive code taken directly from the class, just to see how well the CNN performs, without paying attention to the model size. We just take the MNIST data and convert to Tensor and normalize it and added random rotation from -7 degrees to +7 degrees. It has close to 13,808 model params which are 1.73X our ideal size, but we get a max test accuracy of 99.44 in 13th epoch (but dips down by 15th epoch). The basic backbone is input block -> 1 conv blocks (conv + ReLU) -> transition block (2x2 maxpool) -> 4 conv blocks (conv + ReLU) -> GAP layer -> final conv2d layer. We set the LR to 0.01 with SGD, and a StepLR scheduler with step size 6 and gamma 0.5.